### PR TITLE
feat: computably extend a function from `Fin n` to a function from `Nat`

### PR DESCRIPTION
**Pull request by alexkeizer**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 1]
```diff
@@ -1007,4 +1007,43 @@ theorem sigma_eq_iff_eq_comp_cast {α : Type*} {a b : Σii, Fin ii → α} :
...
+section ExtendFun
+
+/-- Extend a function `f : Fin n → α` to a function on all natural numbers, by
+    defining `f i = a` for all `i ≥ n` and some given constant `a` -/
+def extendFun {α : Type*} {n: ℕ} (f : Fin n → α) (a : α) : ℕ → α :=
+  fun i =>
+    if h : i < n then
+      f ⟨i, h⟩
+    else
+      a
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1019]: This is just
```suggestion
def extendFun {α : Type*} {n: ℕ} (f : Fin n → α) (a : α) : ℕ → α :=
  Function.extend Fin.val f n
```
so in fact I don't think we should add a new definition.

**Review by YaelDillies — COMMENTED** [review]
If you don't manage to use `Function.extend`, please come back to me.

But long story short: Everything that you want to do with this PR is already doable with current mathlib.

File: Mathlib/Data/Fin/Tuple/Basic.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/Fin/Tuple/Basic.lean:1019]: Unfortunately `Function.extend` is noncomputable, so I don't think that's going to work for @alexkeizer. Perhaps we can use `Set.piecewise` instead?

Of course, I'd prefer we work out what makes the `Fin` version of the function so hard to use in #5920

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 2]
```diff
@@ -1007,4 +1007,39 @@ theorem sigma_eq_iff_eq_comp_cast {α : Type*} {a b : Σii, Fin ii → α} :
     sigma_eq_of_eq_comp_cast _ h'⟩
 #align fin.sigma_eq_iff_eq_comp_cast Fin.sigma_eq_iff_eq_comp_cast
 
+section ExtendFun
+
+/-- Extend a function `f : Fin n → α` to a function on all natural numbers, by
+    defining `f i = a` for all `i ≥ n` and some given constant `a` -/
+abbrev extendFun {α : Type*} {n : ℕ} (f : Fin n → α) (a : α) : ℕ → α :=
+  Set.piecewiseMem (· < n) (fun i h => f ⟨i, h⟩) (fun _ _ => a)
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1015]:
```suggestion
  Set.piecewiseMem {x | x < n} (fun i h => f ⟨i, h⟩) (fun _ _ => a)
```

File: Mathlib/Data/Fin/Tuple/Basic.lean (hunk 1 above)
Comment (alexkeizer) [reply Mathlib/Data/Fin/Tuple/Basic.lean]: I've generalize `Set.piecewise` to 
```lean
def Set.piecewiseMem {α : Type u} {β : α → Sort v} (s : Set α) [DecidablePred (· ∈ s)]
    (f : ∀ i ∈ s, β i) (g : ∀ i ∉ s, β i) : ∀ i, β i :=
  fun i ↦ if h : i ∈ s then f i h  else g i h
```

And used this to define `Fin.extendFun`.

File: Mathlib/Init/Set.lean [hunk 3]
```diff
@@ -51,6 +51,11 @@ protected def Mem (a : α) (s : Set α) : Prop :=
 instance : Membership α (Set α) :=
   ⟨Set.Mem⟩
 
+/-- A `Set` is just a predicate, so membership of the set is decidable whenever the predicate is
+    decidable -/
+instance {α} (s : Set α) [DecidablePred s] : DecidablePred (· ∈ s) :=
+  inferInstanceAs (DecidablePred s)
+
```
[the remark this answers is not in the document]
Comment (alexkeizer) [reply Mathlib/Init/Set.lean:58]: Indeed, I should have used the `setOf` notation. Thanks for pointing that out, I've removed the instance

File: Mathlib/Data/Fin/Tuple/Basic.lean (hunk 1 above)
Comment (tobiasgrosser) [reply Mathlib/Data/Fin/Tuple/Basic.lean]: I fixed the build, so all is green now.

**Comment by alexkeizer** [comment]
This PR has changed a bit.
I've generalized `Set.piecewise` to a definition `Set.piecewiseMem` where the constituent functions may also depend on the (non-)membership of its argument. Then, I've used this to define a more generic `Subtype.extendFun`, which (computably) extends a function from a subtype to the original type.
However, some of the lemmas that I needed for the original PR are really `Fin` specific, so I've kept `Fin.extendFun` as an instantiation of `Subtype.extendFun`. I guess I could just inline this definition in the lemmas, if it's not desirable to have a new definition.

@YaelDillies it would be good to get your thoughts, are you still opposed to a new definition, with the extra context that we need it to be computable?

**Comment by YaelDillies** [comment]
Just noting that you should merge master to fix the current conflicts and that I will unavailable to review this for another week.

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 4]
```diff
@@ -1065,4 +1066,32 @@ theorem sigma_eq_iff_eq_comp_cast {α : Type*} {a b : Σii, Fin ii → α} :
...
+    extendFun f a i.val = f (i.castAdd _) := by
+  apply Subtype.extendFun_of_p
+  apply Nat.lt_add_right _ _ _ i.isLt
+
+@[simp] lemma extendFun_comp_val' {α a n} (m) {f : Fin (n+m) → α} :
+    (extendFun f a) ∘ val = f ∘ (castAdd m) := by
+  funext i; apply extendFun_val'
+
+@[simp] lemma extendFun_add {α a n} (m) {f : Fin (n+(m+1)) → α} :
+    (extendFun f a) n = f ⟨n, by simp⟩ := by
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1087]: The statement of this one looks rather strange to me; I think it should instead be:
```
@[simp] lemma extendFun_apply_of_le {n} (f : Fin m → α) (a : α) (n : ℕ) (hn : n < m) :
    extendFun f a n = f ⟨n, hn⟩ := by
 ```

File: Mathlib/Data/Fin/Tuple/Basic.lean (hunk 1 above)
Comment (tobiasgrosser) [reply Mathlib/Data/Fin/Tuple/Basic.lean]: This has now been implemented by Alex.

**Comment by urkud** [comment]
What are the use cases? Do you want your definition to be computable? If not, then you can use `Function.extend`.

**Comment by alexkeizer** [comment]
Indeed, the original use-case requires computability, see also the discussion in @YaelDillies earlier comment that brings up `Function.extend`. I've added some more documentation to emphasize this point.

File: Mathlib/Data/Subtype/ExtendFun.lean [hunk 5]
```diff
@@ -14,10 +14,26 @@ namespace Subtype
 
 variable {α β : Type*} {p : α → Prop} [DecidablePred p]
 
-/-- Extend a function on a subtype of `α` to a function on `α`. -/
+/-- Extend a function on a subtype of `α` to a function on `α`.
+This is similar to `Function.extend`, but differs on two main points: it's computable, and
```
Comment (YaelDillies) [inline Mathlib/Data/Subtype/ExtendFun.lean:18]:
```suggestion

This is similar to `Function.extend`, but differs on two main points: it's computable, and
```

**Comment by bollu** [comment]
I did another round of nit-cleaning. From my view, this PR is mergeable.

**Review by YaelDillies — COMMENTED** [review]
I'm still not a huge fan of "add as many specialisations of `ite` as we can think of". Special cases are good when they are motivated, but here you have two new definitions which are basically just wrappers around `ite` with little to no discussion about the different choices you have made. Eg
* Why you make one function go from a subtype and the other one a dependent function
* Why you don't allow a dependent output
* Why you require `DecidablePred` rather than using `Classical.dec`

If I merge this now, people in the future will do the same as you: add a definition with a docstring explaining why they could not use the existing definitions (now including yours) but failing to explain why they can't just use `ite`/`dite`.

File: Mathlib/Data/Subtype/ExtendFun.lean [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace Subtype
+
+variable {α β : Type*} {p : α → Prop} [DecidablePred p]
+
+/-- Extend a function on a subtype of `α` to a function on `α`.
+
+This is similar to `Function.extend`, but differs on two main points: it's computable, and
+the function `g`, that maps elements of `α` which are *not* in the subtype,
+is given access to a proof of non-membership of its argument in this definition.
```
[the remark this answers is not in the document]
Comment (alexkeizer) [reply Mathlib/Data/Subtype/ExtendFun.lean:21]: Not particularly, I only care about the computability aspect

**Comment by alexkeizer** [comment]
> I'm still not a huge fan of "add as many specialisations of `ite` as we can think of". Special cases are good when they are motivated, but here you have two new definitions which are basically just wrappers around `ite` with little to no discussion about the different choices you have made. Eg
> 
>     * Why you make one function go from a subtype and the other one a dependent function
> 
>     * Why you don't allow a dependent output
> 
>     * Why you require `DecidablePred` rather than using `Classical.dec`
> 
> 
> If I merge this now, people in the future will do the same as you: add a definition with a docstring explaining why they could not use the existing definitions (now including yours) but failing to explain why they can't just use `ite`/`dite`.

To preface this discussion: my immediate need for this functionality has disappeared now that the `ofBits` function got scrapped from #5920. I am happy to put a pin in this for now (if I recall correctly, @mhk119 was the one who originally proposed `ofBits`, so they might want to revive that discussion).

So, the thing I really care(d) about is, as the PR-title reveals, computably extending a function from `Fin n` to `Nat`.
Defining this conversion in terms of `ite` is not too hard, but I also care(d) about some lemmas involving this conversion.
Rather than copy the `ite` term in all these lemmas, it seemed sensible to me to have a name for this conversion: that was the original PR I submitted.

The feedback I got was that this was too specific, there was a desire to relate this to `Set.piecewise`. This doesn't work as-is, as I mentioned in an earlier comment.
> `Set.piecewise` also does not work directly, since the function f doesn't have access to the fact that its argument is in the set, thus it's impossible to construct the appropriate Fin n inhabitant.

Thus, I defined a `Set.piecewiseMem` which is allowed to depend on a proof of membership. In my particular use-case, I don't need the proof of non-membership in the alternative case, but I figured someone might eventually, so I kept things symmetric.

The question of dependent output didn't come to my mind, since I didn't need it, but I'm more than happy to include that generalization as well following the same logic.

`Classical.dec` is not suitable, since it is noncomputable

**Comment by alexkeizer** [comment]
I'm closing this since it's not getting reviewed, and my use-case for it has disappeared

**Pull request closed without merging**